keras_nlp/utils/python_utils_test.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_nlp.utils.python_utils import classproperty
+from keras_nlp.utils.python_utils import format_docstring
+
+
+class ClassWithClassProperty:
+    @classproperty
+    def foo(cls):
+        return "class property"
+
+
+@format_docstring(adjective="salubrious")
+def foo():
+    """It was a {adjective} November day."""
+    return "function"
+
+
+@format_docstring(adjective="smelly", name="Mortimer")
+class ClassWithDocstrings:
+    """I saw my {adjective} friend {name}."""
+
+    def __init__(self):
+        self.bar = "property"
+
+    @classmethod
+    @format_docstring(noun="cactus", bodypart="nostril")
+    def foo(cls):
+        """He was holding a {noun} in his {bodypart}."""
+        return "class method"
+
+
+class ClassPropertyTest(tf.test.TestCase):
+    def test_class_property(self):
+        self.assertAllEqual(ClassWithClassProperty.foo, "class property")
+
+
+class FormatDocstringTest(tf.test.TestCase):
+    def test_function(self):
+        self.assertAllEqual(foo(), "function")
+        self.assertAllEqual(foo.__doc__, "It was a salubrious November day.")
+
+    def test_class(self):
+        self.assertAllEqual(ClassWithDocstrings().bar, "property")
+        self.assertAllEqual(
+            ClassWithDocstrings.__doc__, "I saw my smelly friend Mortimer."
+        )
+
+    def test_class_method(self):
+        self.assertAllEqual(ClassWithDocstrings.foo(), "class method")
+        self.assertAllEqual(
+            ClassWithDocstrings.foo.__doc__,
```

Comment:
Do you know whether `__doc__` or `__func__.__doc__` is used by keras.io?

Reply:
That is a bit of a rabbit hole. Keras.io uses `inspect.getdoc`, which is a core python utility.

Digging into the [source for that](https://github.com/python/cpython/blob/3.11/Lib/inspect.py#L786-L846), it looks like it will ultimately use the first hit in `cls.__mro__` which is a builtin part of python for [resolving methods](https://www.educative.io/answers/what-is-mro-in-python). I'm guessing that will end up being `cls.name.__doc__`? But not sure.

Ultimately, I do not think it matters, as `cls.name.__doc__` is a read only view of the value in `cls.name.__func__.__doc__`.